Configure embedding per Window instead of per Viewport

konkitoman opened the pull request:
_Closes https://github.com/godotengine/godot-proposals/issues/4741_

This adds `Window.native_hint` this makes undocking a window easier.
For that to work the `Viewport.gui_embed_subwindows` needs to be `false`.

Tested only for Arch Linux, KDE Plasma 6.2.3, Wayland

konkitoman committed:
Some simple docs

Calinou commented:
Thanks for opening a pull request :slightly_smiling_face: 

Feature pull requests should be associated to a *feature proposal* to justify the need for implementing the feature in Godot core. Please open a proposal on the [Godot proposals](https://github.com/godotengine/godot-proposals) repository (and link to this pull request in the proposal's body).

reduz commented:
This is quite complex and the use case is not clear, please open a proposal so this feature can be discussed.

Calinou commented:
> This is quite complex and the use case is not clear, please open a proposal so this feature can be discussed.

A proposal was opened at https://github.com/godotengine/godot-proposals/issues/4741.

PoolloverNathan commented:
Pull requests need to be squashed before they are merged.

IceflowRE commented on scene/gui/popup_menu.cpp:
```diff
@@ -181,7 +181,11 @@ void PopupMenu::_activate_submenu(int p_over, bool p_by_keyboard) {
-	Point2 this_pos = get_position();
+	Point2 this_pos = Point2(0, 0);
+
+	if (is_current_embedded())
+		this_pos = get_position();
```
```suggestion
	if (is_current_embedded()) {
		this_pos = get_position();
	}
```

IceflowRE commented on scene/main/viewport.cpp:
```diff
@@ -198,7 +198,8 @@ void Viewport::_sub_window_update(Window *p_window) {
 	}
 
-	ERR_FAIL_COND(index == -1);
+	if (index == -1)
+		return; //ERR_FAIL_COND(index == -1);
```
```suggestion
	if (index == -1) {
		return; //ERR_FAIL_COND(index == -1);
	}
```

IceflowRE commented on scene/main/window.cpp:
```diff
@@ -222,12 +222,129 @@ void Window::set_ime_position(const Point2i &p_pos) {
+	Viewport *_embedder = _get_embedder();
+
+	bool force_embedded = false;
+	if (_embedder != nullptr)
+		force_embedded = _embedder->is_embedding_subwindows();
```
```suggestion
	if (_embedder != nullptr) {
		force_embedded = _embedder->is_embedding_subwindows();
	}
```

IceflowRE commented on scene/main/window.cpp:
```diff
@@ -222,12 +222,129 @@ void Window::set_ime_position(const Point2i &p_pos) {
+}
+
+void Window::set_embedded(bool p_enable) {
+	if (is_inside_tree() && (!Engine::get_singleton()->is_editor_hint()))
+		_destroy_window();
+	embedded = p_enable;
+	if (is_inside_tree() && (!Engine::get_singleton()->is_editor_hint()))
+		_init_window();
```
```suggestion
	if (is_inside_tree() && (!Engine::get_singleton()->is_editor_hint())) {
		_destroy_window();
	}
	embedded = p_enable;
	if (is_inside_tree() && (!Engine::get_singleton()->is_editor_hint())) {
		_init_window();
	}
```

IceflowRE commented on scene/main/viewport.cpp:
```diff
@@ -1790,22 +1792,22 @@ void Viewport::_gui_input_event(Ref<InputEvent> p_event) {
-				if (w) {
-					if (w->is_embedded()) {
-						embedder = w->_get_embedder();
+			//if (is_embedding_subwindows()) {
```
Why was the following commented out?

konkitoman replied:
I think that i was experimenting at that time!

I forgot about this PR!
This currently works?

konkitoman replied:
I updated this PR!

AThousandShips commented on scene/main/window.cpp:
```diff
@@ -1196,6 +1213,109 @@ Viewport *Window::get_embedder() const {
+			_update_window_size();
+		}
+
+	} else {
+		if (!get_parent()) {
+			// It's the root window!
+			visible = true; // Always visible.
+			window_id = DisplayServer::MAIN_WINDOW_ID;
+			DisplayServer::get_singleton()->window_attach_instance_id(get_instance_id(), window_id);
+			_update_from_window();
+			// Since this window already exists (created on start), we must update pos and size from it.
+			{
+				position = DisplayServer::get_singleton()->window_get_position(window_id);
+				size = DisplayServer::get_singleton()->window_get_size(window_id);
+				focused = DisplayServer::get_singleton()->window_is_focused(window_id);
+			}
+			_update_window_size(); // Inform DisplayServer of minimum and maximum size.
+			_update_viewport_size(); // Then feed back to the viewport.
+			_update_window_callbacks();
+			RS::get_singleton()->viewport_set_update_mode(get_viewport_rid(), RS::VIEWPORT_UPDATE_WHEN_VISIBLE);
+		} else {
+			// Create.
+			if (visible) {
+				_make_window();
+			}
+		}
+	}
```
```suggestion
	} else if (!get_parent()) {
		// It's the root window!
		visible = true; // Always visible.
		window_id = DisplayServer::MAIN_WINDOW_ID;
		DisplayServer::get_singleton()->window_attach_instance_id(get_instance_id(), window_id);
		_update_from_window();
		// Since this window already exists (created on start), we must update pos and size from it.
		position = DisplayServer::get_singleton()->window_get_position(window_id);
		size = DisplayServer::get_singleton()->window_get_size(window_id);
		focused = DisplayServer::get_singleton()->window_is_focused(window_id);
		_update_window_size(); // Inform DisplayServer of minimum and maximum size.
		_update_viewport_size(); // Then feed back to the viewport.
		_update_window_callbacks();
		RS::get_singleton()->viewport_set_update_mode(get_viewport_rid(), RS::VIEWPORT_UPDATE_WHEN_VISIBLE);
	} else if (visible) {
		// Create.
		_make_window();
	}
```
No need for a nested block here, hurts readability

AThousandShips commented on scene/main/window.cpp:
```diff
@@ -1196,6 +1213,109 @@ Viewport *Window::get_embedder() const {
+		if (embedder) {
+			if (embedder->is_embedding_subwindows()) {
+				embedded = true;
+			} else {
+				if (native) {
+					embedder = nullptr;
+				} else {
+					embedded = true;
+				}
+			}
```
```suggestion
			} else if (native) {
				embedder = nullptr;
			} else {
				embedded = true;
			}
```
Same here

AThousandShips commented on scene/main/window.cpp:
```diff
@@ -807,6 +821,9 @@ void Window::set_visible(bool p_visible) {
 	updating_child_controls = false;
 
 	Viewport *embedder_vp = get_embedder();
+	if (embedder_vp && !(embedder_vp->is_embedding_subwindows() || !native)) {
```
```suggestion
	if (embedder_vp && !embedder_vp->is_embedding_subwindows() && native) {
```
Simpler and equivalent

IceflowRE commented:
We would also require a workflow approval? 🤔

AThousandShips commented:
I'll approve once changes are applied, no reason to run twice 🙂

Geometror commented:
This needs to be rebased and retested.

Since I'd like to move this forward, please let me know if you have the time to work on it. If not, I can take it over and add you as a co-author.

konkitoman commented:
I will try to rebase and test, now.

konkitoman committed:
Added Window.native_hint
Now when Viewport.gui_embedd_subwindows is false, the windows can be embedded by setting native_hint to false.
This makes easier to undock a window.

Geometror commented:
Hmm, I guess parts of this PR can be salvaged in order to implement a `force_embedded` flag (which would be consistent with  `force_native`). This would be very helpful for certain editor popups/dialogs to feel snappier.

Geometror commented:
After taking a look at the code I'd suggest closing this PR since the `force_native` flag superseded the functionality here. Implementing a `force_embedded` flag should be done in a new PR.
